File: jtms/go/jsrules.go
// Basic rules for JSAINT.
// AND/OR subgoal management rules.
// Translated from jsrules.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.

package jtms

// RegisterJSaintRules registers the bookkeeping rules for JSAINT.
// These rules manage AND/OR subgoal expansion, solution detection,
// and failure propagation.
// Corresponds to the rules defined in jsrules.lisp.
func RegisterJSaintRules(jtre *Jtre) {
	// Rule: When AND-SUBGOALS is IN, expand pointers
	andDbclass := JGetDbclass([]interface{}{"AND-SUBGOALS"}, jtre)
	InsertJRule(andDbclass,
		func(p interface{}) (bool, []interface{}, bool) {
			ps, ok := p.([]interface{})
			if !ok || len(ps) < 3 || ps[0] != "AND-SUBGOALS" {
				return false, nil, false
			}
			return true, []interface{}{ps[1], ps[2]}, true
		},
		func(args ...interface{}) {
			if len(args) < 3 {
				return
			}
			parent := args[1]
			children, ok := args[2].([]interface{})
			if !ok {
				return
			}
			for _, child := range children {
				JAssert([]interface{}{"PARENT-OF", child, parent, "AND"},
					[]interface{}{":DEF-OF-AND"}, jtre)
			}
			// Assert solved when all children solved
			solvedJusts := make([]interface{}, 0, len(children)+1)
			solvedJusts = append(solvedJusts, ":AND-SUCCESS")
			for _, child := range children {
				solvedJusts = append(solvedJusts, []interface{}{"SOLVED", child})
			}
			JAssert([]interface{}{"solved", parent}, solvedJusts, jtre)
		},
	)

	// Rule: When OR-SUBGOALS is IN, set up OR expansion
	orDbclass := JGetDbclass([]interface{}{"OR-SUBGOALS"}, jtre)
	InsertJRule(orDbclass,
		func(p interface{}) (bool, []interface{}, bool) {
			ps, ok := p.([]interface{})
			if !ok || len(ps) < 3 || ps[0] != "OR-SUBGOALS" {
				return false, nil, false
			}
			return true, []interface{}{ps[1], ps[2]}, true
		},
		func(args ...interface{}) {
			if len(args) < 3 {
				return
			}
			parent := args[1]
			children, ok := args[2].([]interface{})
			if !ok {
				return
			}
			for _, child := range children {
				JAssert([]interface{}{"PARENT-OF", child, parent, "OR"},
					[]interface{}{":DEF-OF-OR"}, jtre)
			}
		},
	)

	// Rule: When PARENT-OF is IN, mark child as relevant
	parentDbclass := JGetDbclass([]interface{}{"PARENT-OF"}, jtre)
	InsertJRule(parentDbclass,
		func(p interface{}) (bool, []interface{}, bool) {
			ps, ok := p.([]interface{})
			if !ok || len(ps) < 4 || ps[0] != "PARENT-OF" {
				return false, nil, false
			}
			return true, []interface{}{ps[1], ps[2], ps[3]}, true
		},
		func(args ...interface{}) {
			if len(args) < 4 {
				return
			}
			child := args[1]
			JAssert([]interface{}{"RELEVANT", child},
				[]interface{}{":STILL-WORKING-ON"}, jtre)
		},
	)

	// Rule: When SOLUTION-OF is IN, mark as SOLVED
	solnDbclass := JGetDbclass([]interface{}{"SOLUTION-OF"}, jtre)
	InsertJRule(solnDbclass,
		func(p interface{}) (bool, []interface{}, bool) {
			ps, ok := p.([]interface{})
			if !ok || len(ps) < 3 || ps[0] != "SOLUTION-OF" {
				return false, nil, false
			}
			return true, []interface{}{ps[1], ps[2]}, true
		},
		func(args ...interface{}) {
			if len(args) < 3 {
				return
			}
			problem := args[1]
			JAssert([]interface{}{"SOLVED", problem},
				[]interface{}{":FOUND-ANSWER"}, jtre)
		},
	)

	// Rule: When SOLVED, close the problem
	solvedDbclass := JGetDbclass([]interface{}{"SOLVED"}, jtre)
	InsertJRule(solvedDbclass,
		func(p interface{}) (bool, []interface{}, bool) {
			ps, ok := p.([]interface{})
			if !ok || len(ps) < 2 || ps[0] != "SOLVED" {
				return false, nil, false
			}
			return true, []interface{}{ps[1]}, true
		},
		func(args ...interface{}) {
			if len(args) < 2 {
				return
			}
			problem := args[1]
			JRetract([]interface{}{"OPEN", problem}, "EXPAND-AGENDA-ITEM", true, jtre)
		},
	)

	// Rule: When FAILED, close the problem
	failedDbclass := JGetDbclass([]interface{}{"FAILED"}, jtre)
	InsertJRule(failedDbclass,
		func(p interface{}) (bool, []interface{}, bool) {
			ps, ok := p.([]interface{})
			if !ok || len(ps) < 2 || ps[0] != "FAILED" {
				return false, nil, false
			}
			return true, []interface{}{ps[1]}, true
		},
		func(args ...interface{}) {
			if len(args) < 2 {
				return
			}
			problem := args[1]
			JRetract([]interface{}{"OPEN", problem}, "EXPAND-AGENDA-ITEM", true, jtre)
		},
	)
}

File: jtms/java/JSRules.java
// Bookkeeping rules for JSAINT.
// AND/OR subgoal management.
// Translated from jsrules.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.
package jtms;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * JSRules registers the bookkeeping rules for JSAINT:
 * AND/OR subgoal expansion, solution detection, and failure propagation.
 */
public class JSRules {

    /** Register all JSAINT bookkeeping rules. */
    @SuppressWarnings("unchecked")
    public static void registerRules(Jtre jtre) {
        // Rule: AND-SUBGOALS expansion
        Dbclass andDbclass = JData.getDbclass(Arrays.asList("AND-SUBGOALS"), jtre);
        JRules.insertRule(andDbclass,
            p -> {
                List<Object> ps = asList(p);
                if (ps == null || ps.size() < 3 || !"AND-SUBGOALS".equals(ps.get(0)))
                    return new Rule.MatchResult(false, null, false);
                List<Object> bindings = new ArrayList<>();
                bindings.add(ps.get(1));
                bindings.add(ps.get(2));
                return new Rule.MatchResult(true, bindings, true);
            },
            args -> {
                if (args.size() < 3) return;
                Object parent = args.get(1);
                List<Object> children = asList(args.get(2));
                if (children == null) return;
                for (Object child : children) {
                    JData.jassert(Arrays.asList("PARENT-OF", child, parent, "AND"),
                        Arrays.asList(":DEF-OF-AND"), jtre);
                }
            });

        // Rule: OR-SUBGOALS expansion
        Dbclass orDbclass = JData.getDbclass(Arrays.asList("OR-SUBGOALS"), jtre);
        JRules.insertRule(orDbclass,
            p -> {
                List<Object> ps = asList(p);
                if (ps == null || ps.size() < 3 || !"OR-SUBGOALS".equals(ps.get(0)))
                    return new Rule.MatchResult(false, null, false);
                List<Object> bindings = new ArrayList<>();
                bindings.add(ps.get(1));
                bindings.add(ps.get(2));
                return new Rule.MatchResult(true, bindings, true);
            },
            args -> {
                if (args.size() < 3) return;
                Object parent = args.get(1);
                List<Object> children = asList(args.get(2));
                if (children == null) return;
                for (Object child : children) {
                    JData.jassert(Arrays.asList("PARENT-OF", child, parent, "OR"),
                        Arrays.asList(":DEF-OF-OR"), jtre);
                }
            });

        // Rule: PARENT-OF -> RELEVANT
        Dbclass parentDbclass = JData.getDbclass(Arrays.asList("PARENT-OF"), jtre);
        JRules.insertRule(parentDbclass,
            p -> {
                List<Object> ps = asList(p);
                if (ps == null || ps.size() < 4 || !"PARENT-OF".equals(ps.get(0)))
                    return new Rule.MatchResult(false, null, false);
                List<Object> bindings = new ArrayList<>();
                bindings.add(ps.get(1));
                return new Rule.MatchResult(true, bindings, true);
            },
            args -> {
                if (args.size() < 2) return;
                Object child = args.get(1);
                JData.jassert(Arrays.asList("RELEVANT", child),
                    Arrays.asList(":STILL-WORKING-ON"), jtre);
            });

        // Rule: SOLUTION-OF -> SOLVED
        Dbclass solnDbclass = JData.getDbclass(Arrays.asList("SOLUTION-OF"), jtre);
        JRules.insertRule(solnDbclass,
            p -> {
                List<Object> ps = asList(p);
                if (ps == null || ps.size() < 3 || !"SOLUTION-OF".equals(ps.get(0)))
                    return new Rule.MatchResult(false, null, false);
                List<Object> bindings = new ArrayList<>();
                bindings.add(ps.get(1));
                return new Rule.MatchResult(true, bindings, true);
            },
            args -> {
                if (args.size() < 2) return;
                JData.jassert(Arrays.asList("SOLVED", args.get(1)),
                    Arrays.asList(":FOUND-ANSWER"), jtre);
            });

        // Rule: SOLVED -> close problem
        Dbclass solvedDbclass = JData.getDbclass(Arrays.asList("SOLVED"), jtre);
        JRules.insertRule(solvedDbclass,
            p -> {
                List<Object> ps = asList(p);
                if (ps == null || ps.size() < 2 || !"SOLVED".equals(ps.get(0)))
                    return new Rule.MatchResult(false, null, false);
                List<Object> bindings = new ArrayList<>();
                bindings.add(ps.get(1));
                return new Rule.MatchResult(true, bindings, true);
            },
            args -> {
                if (args.size() < 2) return;
                JData.jretract(Arrays.asList("OPEN", args.get(1)),
                    "EXPAND-AGENDA-ITEM", true, jtre);
            });

        // Rule: FAILED -> close problem
        Dbclass failedDbclass = JData.getDbclass(Arrays.asList("FAILED"), jtre);
        JRules.insertRule(failedDbclass,
            p -> {
                List<Object> ps = asList(p);
                if (ps == null || ps.size() < 2 || !"FAILED".equals(ps.get(0)))
                    return new Rule.MatchResult(false, null, false);
                List<Object> bindings = new ArrayList<>();
                bindings.add(ps.get(1));
                return new Rule.MatchResult(true, bindings, true);
            },
            args -> {
                if (args.size() < 2) return;
                JData.jretract(Arrays.asList("OPEN", args.get(1)),
                    "EXPAND-AGENDA-ITEM", true, jtre);
            });
    }

    @SuppressWarnings("unchecked")
    private static List<Object> asList(Object x) {
        if (x instanceof List) return (List<Object>) x;
        return null;
    }
}
